src/System.IO.FileSystem.AccessControl/src/System/IO/FileSystemAclExtensions.netcoreapp.cs

```diff
@@ -24,5 +36,166 @@ public static void Create(this DirectoryInfo directoryInfo, DirectorySecurity di
+        ///-or-
+        /// <paramref name="options" /> is not a supported enum value.
+        ///-or-
+        /// <paramref name="bufferSize" /> is not a positive number.</exception>
+        /// <exception cref="UnauthorizedAccessException">Access to the path is denied.</exception>
+        /// <exception cref="ArgumentException">Invalid handle.
+        ///-or-
+        /// Handle does not support asynchronous operations. The parameters to the <see cref="FileStream" /> constructor may need to be changed to indicate that the handle was opened synchronously (that is, it was not opened for overlapped I/O).
+        /// </exception>
+        /// <exception cref="ObjectDisposedException">Cannot access a closed file.</exception>
+        public static FileStream Create(this FileInfo fileInfo, FileMode mode, FileSystemRights rights, FileShare share, int bufferSize, FileOptions options, FileSecurity fileSecurity)
+        {
+            if (fileInfo == null)
+            {
+                throw new ArgumentNullException(nameof(fileInfo));
+            }
+
+            if (fileSecurity == null)
+            {
+                throw new ArgumentNullException(nameof(fileSecurity));
+            }
+
+            // don't include inheritable in our bounds check for share
+            FileShare tempshare = share & ~FileShare.Inheritable;
+
+            string badArg = null;
+            if (mode < FileMode.CreateNew || mode > FileMode.Append)
+            {
+                badArg = nameof(mode);
+            }
+            else if (tempshare < FileShare.None || tempshare > (FileShare.ReadWrite | FileShare.Delete))
+            {
+                badArg = nameof(share);
+            }
+            if (badArg != null)
+            {
+                throw new ArgumentOutOfRangeException(badArg, SR.ArgumentOutOfRange_Enum);
+            }
+
+            // We need to check only the supported attributes
+            // NOTE: any change to FileOptions enum needs to be matched here in the error validation
+            if (options != FileOptions.None &&
+                (options & ~(FileOptions.WriteThrough | FileOptions.Asynchronous | FileOptions.RandomAccess | FileOptions.DeleteOnClose | FileOptions.SequentialScan | FileOptions.Encrypted | (FileOptions)0x20000000 /* NoBuffering */)) != 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(options), SR.ArgumentOutOfRange_Enum);
+            }
+
+            if (bufferSize <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), SR.ArgumentOutOfRange_NeedPosNum);
+            }
+
+            SafeFileHandle handle = null;
+            try
+            {
+                handle = CreateFileOpenHandle(fileInfo.FullName, mode, rights, share, options, fileSecurity);
+                return new FileStream(handle, GetFileStreamFileAccess(rights), bufferSize, (options & FileOptions.Asynchronous) != 0);
```

Comment:
@JeremyKuhne 
In the original code, we only put the call to the `FileStream` constructor inside the try.
The reason why I put both `CreateFileOpenHandle` and `FileStream` inside the try, is because `Dispose` was not being called when the handle was invalid. Was this really necessary, or should I take the `CreateFileOpenHandle` call out of the try?

Reply:
Resolved after applying Dan's suggestion above.